src/lsm/cache_map.zig

```diff
@@ -0,0 +1,358 @@
+        // 2. After an upsert, we haven't evicted anything, check our stash:
+        //    a. If a matching item exists there, it means we're doing an update of an item that's
+        //       in the stash. Store the original item without clobbering in our scope_map.
+        //    b. If no matching item exists there, it means it's an insert. Store a tombstone
+        //       without clobbering in our scope_map.
+        // 3. After an upsert, we have evicted an item that was not an exact match. This means we're
+        //    doing an insert of a new value, but two keys have the same tags. Store the evicted
+        //    item without clobbering in our scope_map
+        scope_is_active: bool = false,
+        scope_map: Map,
+
+        last_upsert_was_update_with_eviction: bool = undefined,
+        options: CacheMapOptions,
+
+        pub fn init(allocator: std.mem.Allocator, options: CacheMapOptions) !Self {
+            var cache: Cache = try Cache.init(
+                allocator,
+                options.cache_value_count_max,
+                .{ .name = options.name },
+            );
+            errdefer cache.deinit(allocator);
+
+            var map_1: Map = .{};
+            try map_1.ensureTotalCapacity(allocator, options.map_value_count_max);
+            errdefer map_1.deinit(allocator);
+
+            var map_2: Map = .{};
+            try map_2.ensureTotalCapacity(allocator, options.map_value_count_max);
+            errdefer map_2.deinit(allocator);
+
+            var scope_map: Map = .{};
+            try scope_map.ensureTotalCapacity(allocator, options.scope_value_count_max);
+            errdefer scope_map.deinit(allocator);
+
+            return Self{
+                .cache = cache,
+                .map_1 = map_1,
+                .map_2 = map_2,
+                .scope_map = scope_map,
+                .options = options,
+            };
+        }
+
+        pub fn deinit(self: *Self, allocator: std.mem.Allocator) void {
+            self.cache.deinit(allocator);
+            self.map_1.deinit(allocator);
+            self.map_2.deinit(allocator);
+            self.scope_map.deinit(allocator);
+        }
+
+        pub inline fn has(self: *Self, key: Key) bool {
+            return self.cache.get_index(key) != null or
+                self.map_1.getKeyPtr(tombstone_from_key(key)) != null or
+                self.map_2.getKeyPtr(tombstone_from_key(key)) != null;
+        }
+
+        pub inline fn get(self: *Self, key: Key) ?*Value {
+            return self.cache.get(key) orelse
+                self.map_1.getKeyPtr(tombstone_from_key(key)) orelse
+                self.map_2.getKeyPtr(tombstone_from_key(key));
+        }
+
+        pub fn upsert(self: *Self, value: *const Value) void {
+            self.last_upsert_was_update_with_eviction = false;
+            _ = self.cache.upsert_index(value, on_eviction);
+
+            if (self.scope_is_active and !self.last_upsert_was_update_with_eviction) {
```

Comment:
> Also, maybe replace with updated boolean with a enum { insert, update } to make it impossible to get backwards.

:+1: 

> Could the eviction be handled by returning the value (or a ?struct { value, updated, index }) from upsert_index rather than using a callback? It might make the control flow more linear.

The reasoning behind the callback style (and I do agree it's harder to follow) was to avoid an extra branch, since it already gets taken inside the SAC. The result is pretty significant (numbers from 10 accounts, sequential IDs, tmpfs):

* Linear flow, `if (result.evicted != null)` in `fn upsert`: 466k TPS
* Callback flow: 515k TPS